coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,181 @@
+package workspaceusage
+
+import (
+	"context"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+)
+
+var DefaultFlushInterval = 60 * time.Second
+
+// Store is a subset of database.Store
+type Store interface {
+	BatchUpdateWorkspaceLastUsedAt(context.Context, database.BatchUpdateWorkspaceLastUsedAtParams) error
+}
+
+// Tracker tracks and de-bounces updates to workspace usage activity.
+// It keeps an internal map of workspace IDs that have been used and
+// periodically flushes this to its configured Store.
+type Tracker struct {
+	log      slog.Logger            // you know, for logs
+	mut      sync.Mutex             // protects m
+	m        map[uuid.UUID]struct{} // stores workspace ids
+	s        Store                  // for flushing data
+	tickCh   <-chan time.Time       // controls flush interval
+	stopTick func()                 // stops flushing
+	stopCh   chan struct{}          // signals us to stop
+	stopOnce sync.Once              // because you only stop once
+	doneCh   chan struct{}          // signifies that we have stopped
+	flushCh  chan int               // used for testing.
+}
+
+// New returns a new Tracker. It is the caller's responsibility
+// to call Close().
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        make(map[uuid.UUID]struct{}, 0),
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only.
+func WithFlushChannel(c chan int) Option {
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
```

Comment:
Can we really protect this? I say this because if you pass this, you can spawn go routines that infinitely block. 

```suggestion
func WithFlushChannel(c chan int) Option {
	if flag.Lookup("test.v") == nil {
		panic("Developer Error: WithFlushChannel is only for testing, do not use this in production!")
	}
	return func(h *Tracker) {
		h.flushCh = c
	}
}
```

Reply:
Yeah good suggestion. I don't want those being used outside of tests.